### PR TITLE
VSIX rework

### LINKED ISSUE #4: Support VS2013 for CC Editor Extensions
Editor Extensions throw an exception when only VS2013 is installed on a machine. Please provide a new extension that supports VS2013.

Even when registering the missing VS2012 dlls in the GAC the intellisense support to show contracts doesn't work.

### PR DESCRIPTION
**Pull request by sharwell**
### Summary of changes

* Refactor the Visual Studio extension so a single extension works in Visual Studio 2010+
* Fully support Visual Studio 2010-2015 language services
* ~~Visual Studio 2015 support is a placeholder - the assembly exists but the service factory throws `NotSupportedException`~~ (this was completed)

Fixes #4

### Work left (should now be complete)

* ~~This is a preview branch which does not include the changes in #61. #61 needs to be merged first, and then I will update this pull request (I have a branch in my fork that's already prepared to do this efficiently).~~
* ~~Some files are missing copyright headers~~
  * :question: What is the policy on this? Ideally we could just use the [two-line comment from corefx](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Concurrent/src/System/Collections/Concurrent/IProducerConsumerCollection.cs#L1-L2)
* ~~I forgot to add `[assembly: ProvideBindingPath]` (just remembered while writing this)~~
* ~~Update the build script. **buildMSI10.xml** and/or **Setup10.proj** will need major work~~

### CODE REVIEW

**Comment by sharwell** [comment]
@SergeyTeplyakov The copyright headers question is for you and/or Mike and/or similar.

@hubuk Do you want to see this merged before the setup scripts are updated for it, or afterwards?

Aside from the `[ProvideBindingPath]` issue which is straightforward for me to correct, I don't have any immediate plans to change this pull request until I get more feedback about the direction. :smile:

**Comment by sharwell** [comment]
:thought_balloon: The `[ProvideBindingPath]` attribute likely eliminates much of the need for using ILMerge on the contents of the VSIX.

**Comment by mike-barnett** [comment]
I'm fine with using that copyright header.

**Comment by sharwell** [comment]
@hubuk This is now complete (i.e. ready for review) aside from the required changes to the installer. How would you like to proceed?

**Comment by hubuk** [comment]
@sharwell I will be reviewing all your changes, but it may take a wile. I will comment any questions here and mark a changsets as reviewed by submitting +1 on each individually. Do you think that this will help to organize review better?
And regarding a merge. I think that your PR may goes first. I will prepare installer changes by tomorrow.

**Comment by billings7** [comment]
@hubuk re: the installer changes, would those have any effect on or benefit from PR #36 at all?

**Comment by hubuk** [comment]
@billings7 No, my changes only contain a new version update mechanism in buildMSI10.xml
Regarding #36:
The change you introduced worked fine. Unfortunately now it is marked as conflicting and I think it requires some more changes. In addition it would be better not to commit CodeTools binaries into git but build them and use compiled image instead. If you can take a look into this it would be great. If no, I can recreate your changes and align them in a new PR when I finish my current work.

**Comment by sharwell** [comment]
@hubuk @billings7 Pull requests created prior to #61 will show merge conflicts in the UI, but in many cases will not actually have merge conflicts when the merge is run from the command line (and/or the merge conflicts are automatically resolved).

**Comment by billings7** [comment]
@hubuk I will have a look into that when I can. Changing how CodeTools is built would probably be better in a different PR, from what I can remember some part of it needs VS2012 to build it.

**Comment by sharwell** [comment]
I added 3 commits which are the first part of editor support for VS 2015. Locally I only have 2 more methods to implement (`GetAnyCallNodeAboveTriggerPoint` and `GetTargetAtTriggerPoint`) before I can test the editor extensions within Visual Studio 2015.

**Comment by SergeyTeplyakov** [comment]
Does it means that the integration with VS2015 itself is working fine already (except the async part)?

**Comment by sharwell** [comment]
This pull request only deals with the editor extensions. The other compatibility issues for VS 2015 are actually not due to the IDE itself, but rather the new build tools (MSBuild 14.0 and Roslyn-related differences in the IL).

**Comment by SergeyTeplyakov** [comment]
Does it mean that #36 is all that we need for VS2015 support?
I already fixed issues with differentce in the IL in #74 (at least, everything that I know about this issue). If you have anything related to roslyn-based behavior, could you shared it with me?

File: Microsoft.Research/ContractAdornments/CSharp.Roslyn/Intellisense/IntellisenseContractsHelper.cs [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+        }
+
+        var declNode = nodeInQuestion as MemberDeclarationSyntax;
+        if (declNode != null)
+        {
+            if (lastNode == leafNode) { return declNode; }
+            return null;
+        }
+
+#warning What exactly is this doing?
```
Comment (sharwell) [inline Microsoft.Research/ContractAdornments/CSharp.Roslyn/Intellisense/IntellisenseContractsHelper.cs]: :memo: Someone will need to figure out what this was doing, but it didn't seem to be essential for the initial pull request. I am in favor of creating a new issue to resolve this warning after this pull request is merged.

**Pull request merged**